runtime/near-vm-logic/src/alt_bn128.rs

```diff
@@ -0,0 +1,354 @@
+use bn::arith::U256;
+use bn::{pairing_batch, AffineG1, AffineG2, Fq, Fq2, Fr, Group, GroupError, Gt, G1, G2};
+use borsh::{BorshDeserialize, BorshSerialize};
+use std::io::{self, Error, ErrorKind, Write};
+
+use crate::HostError;
+
+const POINT_IS_NOT_ON_THE_CURVE: &str = "point is not on the curve";
+const POINT_IS_NOT_IN_THE_SUBGROUP: &str = "point is not in the subgroup";
+const NOT_IN_FIELD: &str = "integer is not less than modulus";
+
+pub fn ilog2(mut n: u64) -> u64 {
+    assert!(n > 0);
+    let mut r = 0;
+    if n >> 32 != 0 {
+        n >>= 32;
+        r += 32;
+    }
+    if n >> 16 != 0 {
+        n >>= 16;
+        r += 16;
+    }
+    if n >> 8 != 0 {
+        n >>= 8;
+        r += 8;
+    }
+    if n >> 4 != 0 {
+        n >>= 4;
+        r += 4;
+    }
+    if n >> 2 != 0 {
+        n >>= 2;
+        r += 2;
+    }
+    if n >> 1 != 0 {
+        r += 1;
+    }
+    r
+}
+
+pub fn alt_bn128_g1_multiexp_sublinear_complexity_estimate(n_bytes: u64, discount: u64) -> u64 {
+    // details of computation alt_bn128 parameters are available at https://gist.github.com/snjax/90e8b3e7f5c16a983a5f6347d1d28bde
+    const A: u64 = 85158;
+    const B: u64 = 15119;
+    const C: u64 = 682573;
+    const MULTIEXP_ITEM_SIZE: u64 = std::mem::size_of::<(G1, Fr)>() as u64;
+
+    // A+C*n/(log2(n)+4) - B*n - discount
+
+    let n = (n_bytes + MULTIEXP_ITEM_SIZE - 1) / MULTIEXP_ITEM_SIZE;
+    let mut res = A + if n == 0 {
+        0
+    } else {
+        // set linear complexity growth for n > 32768
+        let l = std::cmp::min(ilog2(n), 15);
+        (n * (l + 3) + (1 << (1 + l))) * C / ((l + 4) * (l + 5))
```

Comment:
Could origins of this formula be clarified?

Reply:
complexity formula is A+B*n+C*n/(log2(n)+D)
A, B, C, D were obtained here https://gist.github.com/snjax/90e8b3e7f5c16a983a5f6347d1d28bde

So, your question is about fitting y(n) = n/(log2(n)+4). 

I fit y(n) as piecewise linear function with vertexes (x_i=2^i, y_i=y(x_i)).

So, y_fitted(x) = (x-x_i)*(y_{i+1} - y_i)/(x_{i+1}-x_i)+y_i = 
    (x-2^i)*(2^(i+1)/(i+5)- 2^i/(i+4))/2^i + 2^i/(i+4) =
    ( (x-2^i) * (2 * (i+4) - (i+5)) + 2^i * (i+5)) / (i+4) / (i+5) = 
    ( (x-2^i) * (i+3) + 2^i * (i+5)) / (i+4) / (i+5) = 
    ( x*(i+3) + 2^(i+1) ) / ((i+4) * (i+5))